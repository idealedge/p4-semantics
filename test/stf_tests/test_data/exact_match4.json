{
  "program" : "exact_match4.p4",
  "__meta__" : {
    "version" : [2, 7],
    "compiler" : "https://github.com/p4lang/p4c"
  },
  "header_types" : [
    {
      "name" : "scalars_0",
      "id" : 0,
      "fields" : []
    },
    {
      "name" : "data_t",
      "id" : 1,
      "fields" : [
        ["f4", 128, false],
        ["b1", 8, false],
        ["b2", 8, false]
      ]
    },
    {
      "name" : "data1_t",
      "id" : 2,
      "fields" : [
        ["f1", 128, false]
      ]
    },
    {
      "name" : "data2_t",
      "id" : 3,
      "fields" : [
        ["f2", 128, false]
      ]
    },
    {
      "name" : "data3_t",
      "id" : 4,
      "fields" : [
        ["f3", 128, false]
      ]
    },
    {
      "name" : "standard_metadata",
      "id" : 5,
      "fields" : [
        ["ingress_port", 9, false],
        ["egress_spec", 9, false],
        ["egress_port", 9, false],
        ["clone_spec", 32, false],
        ["instance_type", 32, false],
        ["drop", 1, false],
        ["recirculate_port", 16, false],
        ["packet_length", 32, false],
        ["enq_timestamp", 32, false],
        ["enq_qdepth", 19, false],
        ["deq_timedelta", 32, false],
        ["deq_qdepth", 19, false],
        ["ingress_global_timestamp", 48, false],
        ["lf_field_list", 32, false],
        ["mcast_grp", 16, false],
        ["resubmit_flag", 1, false],
        ["egress_rid", 16, false],
        ["checksum_error", 1, false],
        ["_padding", 4, false]
      ]
    }
  ],
  "headers" : [
    {
      "name" : "scalars",
      "id" : 0,
      "header_type" : "scalars_0",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "standard_metadata",
      "id" : 1,
      "header_type" : "standard_metadata",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "data",
      "id" : 2,
      "header_type" : "data_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "data1",
      "id" : 3,
      "header_type" : "data1_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "data2",
      "id" : 4,
      "header_type" : "data2_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "data3",
      "id" : 5,
      "header_type" : "data3_t",
      "metadata" : false,
      "pi_omit" : true
    }
  ],
  "header_stacks" : [],
  "header_union_types" : [],
  "header_unions" : [],
  "header_union_stacks" : [],
  "field_lists" : [],
  "errors" : [
    ["NoError", 1],
    ["PacketTooShort", 2],
    ["NoMatch", 3],
    ["StackOutOfBounds", 4],
    ["HeaderTooShort", 5],
    ["ParserTimeout", 6]
  ],
  "enums" : [],
  "parsers" : [
    {
      "name" : "parser",
      "id" : 0,
      "init_state" : "start",
      "parse_states" : [
        {
          "name" : "d2",
          "id" : 0,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data2"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : "d3"
            }
          ],
          "transition_key" : []
        },
        {
          "name" : "d3",
          "id" : 1,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data3"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : "more"
            }
          ],
          "transition_key" : []
        },
        {
          "name" : "more",
          "id" : 2,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : []
        },
        {
          "name" : "start",
          "id" : 3,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data1"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : "d2"
            }
          ],
          "transition_key" : []
        }
      ]
    }
  ],
  "deparsers" : [
    {
      "name" : "deparser",
      "id" : 0,
      "order" : ["data1", "data2", "data3", "data"]
    }
  ],
  "meter_arrays" : [],
  "counter_arrays" : [],
  "register_arrays" : [],
  "calculations" : [],
  "learn_lists" : [],
  "actions" : [
    {
      "name" : "NoAction",
      "id" : 0,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "setb1",
      "id" : 1,
      "runtime_data" : [
        {
          "name" : "val",
          "bitwidth" : 8
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["data", "b1"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "exact_match4.p4",
            "line" : 63,
            "column" : 13,
            "source_fragment" : "val, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "exact_match4.p4",
            "line" : 63,
            "column" : 18,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "noop",
      "id" : 2,
      "runtime_data" : [],
      "primitives" : []
    }
  ],
  "pipelines" : [
    {
      "name" : "ingress",
      "id" : 0,
      "init_table" : "test1",
      "tables" : [
        {
          "name" : "test1",
          "id" : 0,
          "source_info" : {
            "filename" : "exact_match4.p4",
            "line" : 68,
            "column" : 0,
            "source_fragment" : "table test1 { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["data1", "f1"],
              "mask" : null
            },
            {
              "match_type" : "exact",
              "target" : ["data2", "f2"],
              "mask" : null
            },
            {
              "match_type" : "exact",
              "target" : ["data3", "f3"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [1, 2, 0],
          "actions" : ["setb1", "noop", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "setb1" : null,
            "noop" : null,
            "NoAction" : null
          },
          "default_entry" : {
            "action_id" : 0,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : []
    },
    {
      "name" : "egress",
      "id" : 1,
      "init_table" : null,
      "tables" : [],
      "action_profiles" : [],
      "conditionals" : []
    }
  ],
  "checksums" : [],
  "force_arith" : [],
  "extern_instances" : [],
  "field_aliases" : [
    [
      "queueing_metadata.enq_timestamp",
      ["standard_metadata", "enq_timestamp"]
    ],
    [
      "queueing_metadata.enq_qdepth",
      ["standard_metadata", "enq_qdepth"]
    ],
    [
      "queueing_metadata.deq_timedelta",
      ["standard_metadata", "deq_timedelta"]
    ],
    [
      "queueing_metadata.deq_qdepth",
      ["standard_metadata", "deq_qdepth"]
    ],
    [
      "intrinsic_metadata.ingress_global_timestamp",
      ["standard_metadata", "ingress_global_timestamp"]
    ],
    [
      "intrinsic_metadata.lf_field_list",
      ["standard_metadata", "lf_field_list"]
    ],
    [
      "intrinsic_metadata.mcast_grp",
      ["standard_metadata", "mcast_grp"]
    ],
    [
      "intrinsic_metadata.resubmit_flag",
      ["standard_metadata", "resubmit_flag"]
    ],
    [
      "intrinsic_metadata.egress_rid",
      ["standard_metadata", "egress_rid"]
    ]
  ]
}