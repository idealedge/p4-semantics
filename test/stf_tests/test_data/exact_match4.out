<T> <k> @nextPacket </k> <headers> <header> <name> %standard_metadata_t </name> <opts> .Map </opts> <fields> egress_spec : 32 ( .FieldMods ) ;  ( egress_spec : 32 ( .FieldMods ) ;  .FieldDecs ) </fields> </header> <header> <name> data_t </name> <opts> "$fixed_width" |-> 144 </opts> <fields> f4 : 128 ( .FieldMods ) ;  ( b1 : 8 ( .FieldMods ) ;  ( b2 : 8 ( .FieldMods ) ;  .FieldDecs ) ) </fields> </header> <header> <name> data3_t </name> <opts> "$fixed_width" |-> 128 </opts> <fields> f3 : 128 ( .FieldMods ) ;  .FieldDecs </fields> </header> <header> <name> data2_t </name> <opts> "$fixed_width" |-> 128 </opts> <fields> f2 : 128 ( .FieldMods ) ;  .FieldDecs </fields> </header> <header> <name> data1_t </name> <opts> "$fixed_width" |-> 128 </opts> <fields> f1 : 128 ( .FieldMods ) ;  .FieldDecs </fields> </header> </headers> <arrays> .Map </arrays> <fieldLists> .FieldListCellBag </fieldLists> <fieldListCalcs> .FieldListCalcCellBag </fieldListCalcs> <calcFields> .CalcFieldCellBag </calcFields> <instances> <instance> <valid> true </valid> <metadata> true </metadata> <typeName> %standard_metadata_t </typeName> <name> standard_metadata </name> <fieldVals> ingress_port |-> @val ( 2 , 32 , false ) egress_spec |-> @val ( 3 , 32 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> false </metadata> <typeName> data_t </typeName> <name> data </name> <fieldVals> f4 |-> @val ( 61084913312720243968763869790979 , 128 , false ) b2 |-> @val ( 102 , 8 , false ) b1 |-> @val ( 7 , 8 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> false </metadata> <typeName> data3_t </typeName> <name> data3 </name> <fieldVals> f3 |-> @val ( 2668881309183831085987251822994260482 , 128 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> false </metadata> <typeName> data2_t </typeName> <name> data2 </name> <fieldVals> f2 |-> @val ( 2668881309183831085987251822994260482 , 128 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> false </metadata> <typeName> data1_t </typeName> <name> data1 </name> <fieldVals> f1 |-> @val ( 2668881309183831085987251822994260482 , 128 , false ) </fieldVals> </instance> </instances> <initInstances> <instances>-fragment <instance> <valid> true </valid> <metadata> true </metadata> <typeName> %standard_metadata_t </typeName> <name> standard_metadata </name> <fieldVals> ingress_port |-> @undef egress_spec |-> @undef </fieldVals> </instance> <instance> <valid> false </valid> <metadata> false </metadata> <typeName> data_t </typeName> <name> data </name> <fieldVals> .Map </fieldVals> </instance> <instance> <valid> false </valid> <metadata> false </metadata> <typeName> data3_t </typeName> <name> data3 </name> <fieldVals> .Map </fieldVals> </instance> <instance> <valid> false </valid> <metadata> false </metadata> <typeName> data2_t </typeName> <name> data2 </name> <fieldVals> .Map </fieldVals> </instance> <instance> <valid> false </valid> <metadata> false </metadata> <typeName> data1_t </typeName> <name> data1 </name> <fieldVals> .Map </fieldVals> </instance> </instances>-fragment </initInstances> <initArrays> .Map </initArrays> <parserStates> <state> <name> more </name> <body> ( extract ( data ) ; .ExtractOrSetStatements ) return ingress ; </body> </state> <state> <name> d3 </name> <body> ( extract ( data3 ) ; .ExtractOrSetStatements ) return more ; </body> </state> <state> <name> d2 </name> <body> ( extract ( data2 ) ; .ExtractOrSetStatements ) return d3 ; </body> </state> <state> <name> start </name> <body> ( extract ( data1 ) ; .ExtractOrSetStatements ) return d2 ; </body> </state> </parserStates> <valueSets> .ValueSetCellBag </valueSets> <exceptionHandlers> .ExceptionHandlerCellBag </exceptionHandlers> <handledExceptions> .Set </handledExceptions> <statefuls> .StatefulCellBag </statefuls> <tableDirects> .Map </tableDirects> <actions> <action> <name> setb1 </name> <params> val , ( port , .ParamList ) </params> <body> modify_field ( ( data . b1 ) , ( val , .Args ) ) ;  ( modify_field ( ( standard_metadata . egress_spec ) , ( port , .Args ) ) ;  .ActionStatements ) </body> </action> <action> <name> noop </name> <params> .ParamList </params> <body> .ActionStatements </body> </action> </actions> <actionProfiles> .ActionProfileCellBag </actionProfiles> <actionSelectors> .ActionSelectorCellBag </actionSelectors> <tables> <table> <name> test1 </name> <reads> ( data1 . f1 ) : exact ;  ( ( data2 . f2 ) : exact ;  ( ( data3 . f3 ) : exact ;  .FieldMatchs ) ) </reads> <acts> actions { setb1 ;  ( noop ;  .ActionNameItems ) } </acts> <opts> .TableOptionals </opts> <rules> ListItem ( $rule ( 0 , $ctr ( ListItem ( @val ( 1334440654591915542993625911497130241 , 0 , false ) ) ListItem ( @val ( 1334440654591915542993625911497130241 , 0 , false ) ) ListItem ( @val ( 1334440654591915542993625911497130241 , 0 , false ) ) ) , @call ( setb1 , $resolved ( ListItem ( @val ( 127 , 0 , false ) ) ListItem ( @val ( 2 , 0 , false ) ) ) ) ) ) ListItem ( $rule ( 1 , $ctr ( ListItem ( @val ( 2668881309183831085987251822994260482 , 0 , false ) ) ListItem ( @val ( 2668881309183831085987251822994260482 , 0 , false ) ) ListItem ( @val ( 2668881309183831085987251822994260482 , 0 , false ) ) ) , @call ( setb1 , $resolved ( ListItem ( @val ( 7 , 0 , false ) ) ListItem ( @val ( 3 , 0 , false ) ) ) ) ) ) </rules> <default> . </default> </table> </tables> <controlFlows> <control> <name> ingress </name> <body> apply ( test1 ) { .HitMissCases }  .ControlStatements </body> </control> </controlFlows> <cfset> SetItem ( ingress ) </cfset> <calcFieldSet> .Set </calcFieldSet> <ctx> <frameStack> .List </frameStack> <crnt> "$rule" |-> . "$table" |-> . "$action" |-> setb1 "$miss" |-> false </crnt> </ctx> <packet> "0000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000100000000000000000000000110000001100000000000000000000001100000011000000000000000000000011000000110000000000000000000000110000001101010101011001100111011110001000" </packet> <packetout> .List </packetout> <parser> <graph> <onpath> .Set </onpath> <marked> <mstates> SetItem ( more ) SetItem ( start ) SetItem ( d2 ) SetItem ( d3 ) </mstates> <mheaders> SetItem ( data1 ) SetItem ( data2 ) SetItem ( data ) SetItem ( data3 ) </mheaders> </marked> <dporder> ListItem ( data1 ) ListItem ( data2 ) ListItem ( data3 ) ListItem ( data ) </dporder> </graph> <pctx> <index> 528 </index> <lastExt> data </lastExt> <varWidth> . </varWidth> <packetSize> @val ( 68 , 0 , false ) </packetSize> </pctx> </parser> <buffer> <in> .List </in> <out> ListItem ( $packet ( ListItem ( @val ( 1334440654591915542993625911497130241 , 128 , false ) ) ListItem ( @val ( 1334440654591915542993625911497130241 , 128 , false ) ) ListItem ( @val ( 1334440654591915542993625911497130241 , 128 , false ) ) ListItem ( @val ( 40723275541813495979175913193986 , 128 , false ) ) ListItem ( @val ( 127 , 8 , false ) ) ListItem ( @val ( 102 , 8 , false ) ) ListItem ( @val ( 30600 , 16 , false ) ) , 2 ) ) ListItem ( $packet ( ListItem ( @val ( 2668881309183831085987251822994260482 , 128 , false ) ) ListItem ( @val ( 2668881309183831085987251822994260482 , 128 , false ) ) ListItem ( @val ( 2668881309183831085987251822994260482 , 128 , false ) ) ListItem ( @val ( 61084913312720243968763869790979 , 128 , false ) ) ListItem ( @val ( 7 , 8 , false ) ) ListItem ( @val ( 102 , 8 , false ) ) ListItem ( @val ( 30600 , 16 , false ) ) , 3 ) ) </out> </buffer> </T>
